staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -572,8 +574,29 @@ func (w *watchCache) Replace(objs []interface{}, resourceVersion string) error {
 	w.Lock()
 	defer w.Unlock()
 
-	w.startIndex = 0
-	w.endIndex = 0
+	// We have to ensure that startIndex is never decreased.
```

Comment:
does `doCacheResizeLocked` have the potential to move `startIndex` backwards with this line:

```
	if capacity < w.capacity {
		// adjust startIndex if cache capacity shrink.
		w.startIndex = w.endIndex - capacity
	}
```

Reply:
This function is only called if `isCacheFullLocked()` is true, which means that w.startIndex+w.capacity == w.endIndex.

So if capacity <w.capacity and the above is true, the startIndex will only move forward.

So we're good here.